crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2951,6 +2985,137 @@ pub enum TypeVarBoundOrConstraints<'db> {
+        self.error_kind.enter_type(db)
+    }
+
+    fn fallback_enter_type(&self, db: &'db dyn Db) -> Type<'db> {
+        self.enter_type(db).unwrap_or(Type::unknown())
+    }
+
+    /// Reports the diagnostic for this error
+    fn report_diagnostic(
+        &self,
+        context: &InferContext<'db>,
+        context_manager_node: ast::AnyNodeRef,
+    ) {
+        self.error_kind
+            .report_diagnostic(context, self.context_manager_type, context_manager_node);
+    }
+}
+
+#[derive(Debug)]
+enum ContextManagerErrorKind<'db> {
+    Exit {
+        enter_ret_ty: Type<'db>,
+        exit_error: CallDunderError<'db>,
+    },
+    Enter(CallDunderError<'db>),
+    EnterAndExit {
+        enter_error: CallDunderError<'db>,
+        exit_error: CallDunderError<'db>,
+    },
+}
+
+impl<'db> ContextManagerErrorKind<'db> {
+    fn enter_type(&self, db: &'db dyn Db) -> Option<Type<'db>> {
+        match self {
+            ContextManagerErrorKind::Exit {
+                enter_ret_ty,
+                exit_error: _,
+            } => Some(*enter_ret_ty),
+            ContextManagerErrorKind::Enter(enter_error)
+            | ContextManagerErrorKind::EnterAndExit {
+                enter_error,
+                exit_error: _,
+            } => match enter_error {
+                CallDunderError::PossiblyUnbound(call_outcome) => {
+                    Some(call_outcome.return_type(db))
+                }
+                CallDunderError::Call(call_error) => call_error.return_type(db),
+                CallDunderError::MethodNotAvailable => None,
+            },
+        }
+    }
+
+    fn format_call_dunder_error(call_dunder_error: &CallDunderError<'db>, name: &str) -> String {
+        match call_dunder_error {
+            CallDunderError::MethodNotAvailable => format!("it doesn't implement `{name}`"),
+            CallDunderError::PossiblyUnbound(_) => {
+                format!("the method `{name}` is possibly unbound")
+            }
+            // TODO: Use more specific error messages for the different error cases.
+            //  E.g. hint toward the union variant that doesn't correctly implement enter,
+            //  distinguish between a not callable `__enter__` attribute and a wrong signature.
+            CallDunderError::Call(_) => format!("it does not correctly implement `{name}`"),
+        }
+    }
+
+    fn format_call_dunder_errors(
+        error_a: &CallDunderError<'db>,
+        name_a: &str,
+        error_b: &CallDunderError<'db>,
+        name_b: &str,
+    ) -> String {
+        match (error_a, error_b) {
+            (CallDunderError::PossiblyUnbound(_), CallDunderError::PossiblyUnbound(_)) => {
+                format!("the methods `{name_a}` and `{name_b}` are possibly unbound")
+            }
+            (CallDunderError::MethodNotAvailable, CallDunderError::MethodNotAvailable) => {
+                format!("it doesn't implement `{name_a}` and `{name_b}`")
```

Comment:
very small nit: we should probably be consistent with respect to "doesn't" vs "does not"

Reply:
I would go with "does not"